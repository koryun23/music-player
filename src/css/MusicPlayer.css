.music-player {
    width: 90%;
    height: 500px;
    margin-bottom: 10px;
    position: absolute;
    top: 20px;
    left: 5%;
    right: 5%;
    display: grid;
    grid-template-columns: auto;
}