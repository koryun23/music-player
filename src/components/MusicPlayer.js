import React, { useEffect } from "react";
import "../css/MusicPlayer.css";
import Toolbar from "./toolbar/Toolbar";
import SongList from "./songs/SongList";
import MusicUploadForm from "./upload/MusicUploadForm";
import { useState } from "react";
import { upload } from "@testing-library/user-event/dist/upload";
import Error from "./Error";
import Success from "./Success";

var currentAudio = null;

export default function MusicPlayer(props) {
    
    const [uploadedFiles, setUploadedFiles] = useState([]);

    const [songList, setSongList] = useState([]);
    const [favoriteSongList, setFavoriteSongList] = useState([]);

    const [currentQueue, setCurrentQueue] = useState([]);
    
    const [filteredSongList, setFilteredSongList] = useState([]);

    const [currentSong, setCurrentSong] = useState("");
    const [isPlaying, setIsPlaying] = useState(false);

    const onPlayAll = (event, mode) => {
        event.preventDefault();
        if(currentQueue.length == 0) {
            onError("There are no songs currently in a queue. Consider adding songs to a queue");
            removeError();
            return;
        }
        let filesToPlay = [];
        console.log(uploadedFiles);
        for(let i = 0; i < uploadedFiles.length; i++) {
            for(let j = 0; j < currentQueue.length; j++) {
                let currentFileName = currentQueue[j].fileName;
                if(uploadedFiles[i].name === currentFileName) {
                    filesToPlay.push(uploadedFiles[i]);
                }
            }
        }
        console.log(filesToPlay);
        if(mode === "once") {
            playAllOnce(filesToPlay, () => console.log(filesToPlay))
        } else if(mode === "repeat") {
            playAllRepeat(filesToPlay, () => console.log(filesToPlay))
        }
    };

    const playAllOnce = (files, cb) => {
        cb();
    }

    const playAllRepeat = (files, cb) => {
        cb();
    }
    
    const onAddAll = (event, mode) => {
        event.preventDefault();
        console.log(mode);
        if(mode === "sequential") {
            addAllSequential(() => console.log(currentQueue));
        } else if(mode === "shuffled"){
            addAllShuffled(() => console.log(currentQueue));
        }
    };

    const addAllSequential = (cb) => {
        let queue = songList.map(song => song);
        setCurrentQueue(queue);
        cb();
    }

    const addAllShuffled = () => {
        let array = songList.map(song => song);
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
        setCurrentQueue(array);
        console.log(array);
    }

    const onFilter = (inputWord) => {
        console.log(inputWord);
        if(!inputWord) {
            setFilteredSongList(songList);
            return;
        }
        let temp = songList.filter(song => (
            song.songName.toLowerCase().includes(inputWord.toLowerCase()) || song.artistName.toLowerCase().includes(inputWord.toLowerCase())
        ));
        console.log(temp);
        setFilteredSongList(temp);
    }

    const onAddSongToPlaylist = (event, songName, artistName, fileName) => {
        event.preventDefault();
        let songListTemp = songList.map(song => song);
        console.log(songListTemp);
        let trackNumber = songListTemp.length + 1;
        let newSong = {
            songName: songName, artistName: artistName, trackNumber: trackNumber, fileName: fileName
        };
        for(let i = 0; i < songListTemp.length; i++) {
            if(songListTemp[i].songName === songName) {
                onError(`There already exists a song named '${songName}' in your playlist: please specify another name`)
                removeError();
                return;
            }
            if(songListTemp[i].fileName === fileName) {
                onError(`There already exists a song imported from the specified file`)
                removeError();
                return;
            }
        }
        songListTemp.push(newSong);
        setSongList(songListTemp);
    }

    const onAddToFavorites = (songName) => {
        let tempFavorites = favoriteSongList.map(song => song);
        let tempAllSongs = songList.map(song => song);

        for(let i = 0; i < tempAllSongs.length; i++) {
            if(songName === tempAllSongs[i].songName) {
                tempFavorites.push(tempAllSongs[i]);
                break;
            }
        }
        setFavoriteSongList(tempFavorites);
    }

    const onRemoveFromFavorites = (songName) => {
        let tempFavorites = favoriteSongList.map(song => song);
        let index = -1;
        for(let i = 0; i < tempFavorites.length; i++) {
            if(tempFavorites[i].songName === songName) {
                index = i;
                break;
            }
        }
        tempFavorites.splice(index, 1);
        setFavoriteSongList(tempFavorites)
    }

    const onStopPlaying = (fileName) => {
        console.log(fileName);
        console.log(currentSong)
        if(!currentAudio) return;
        if(fileName !== currentSong) return;
        currentAudio.pause();
        setIsPlaying(false);
        setCurrentSong(false);
        currentAudio = null;
    }

    const onPlayNextSong = (fileName) => {
        console.log(fileName);
        console.log(currentSong);
        if(!currentAudio) return;
        if(currentSong !== fileName) return;
        if(songList.length <= 1) return;
        let tempUploadedFiles = uploadedFiles.map(file => file);
        let fileToPlay = "";
        for(let i = 0; i < tempUploadedFiles.length; i++) {
            let currentFile = tempUploadedFiles[i];
            if(currentFile.name == fileName) {
                if(i == tempUploadedFiles.length - 1) { // current song is the last one on the playlist
                    fileToPlay = tempUploadedFiles[0];
                } else {
                    fileToPlay = tempUploadedFiles[i + 1];
                }
                break;
            }
        }
        setIsPlaying(false);
        setCurrentSong("");
        currentAudio.pause();
        currentAudio = new Audio(URL.createObjectURL(fileToPlay));
        currentAudio.play();
        setIsPlaying(true);
        setCurrentSong(fileToPlay.name);

    }

    const onPlaySingle = (fileName) => {
        console.log(currentSong, fileName);
        if(currentSong === fileName) {
            if(currentAudio.paused){
                currentAudio.play();
                setIsPlaying(true);
            } else {
                currentAudio.pause();
                setIsPlaying(false);
            }
        } else {
            if(currentAudio) {
                currentAudio.pause();
            }
            setIsPlaying(false);
            for(let i = 0; i < uploadedFiles.length; i++) {
                console.log(uploadedFiles[i]);
                if(uploadedFiles[i].name.includes(fileName)) {
                    currentAudio = new Audio(URL.createObjectURL(uploadedFiles[i]));
                    currentAudio.play();
                    setIsPlaying(true);
                    setCurrentSong(fileName);
                    currentAudio.onended = () => {
                        setCurrentSong("")
                        setIsPlaying(false);
                        console.log("ENDED");
                    }
                }
            }
        }
    }

    const onUploadSong = (event) => {
        if(currentAudio) {
            currentAudio.pause();
            setIsPlaying(false);
        }
        const lastUploadedFile = event.target.files[event.target.files.length - 1];
        pushNewFile(lastUploadedFile);
    }

    const pushNewFile = (lastUploadedFile) => {
        let uploadedFilesTemp = uploadedFiles.map(file => file);
        uploadedFilesTemp.push(lastUploadedFile);
        setUploadedFiles(uploadedFilesTemp);
    }

    const onError = (message) => {
        setError(message);
    }

    const removeError = () => {
        setTimeout(() => {
            setError("")
        }, 6000)
    }

    const [error, setError] = useState();

    useEffect(() => {
        setFilteredSongList(songList);
    }, [songList]);

    return (
        <div className="music-player"> 
            {
                error && <Error errorMessage={error}/>
            }
            
             <Toolbar onPlayAll={(event, mode) => onPlayAll(event, mode)} 
                      onAddAll={(event, mode) => onAddAll(event, mode)}
                      onFilter={onFilter} />
             <SongList songList={filteredSongList} 
                       currentQueue={currentQueue} 
                       onPlaySingle={(fileName) => onPlaySingle(fileName)} 
                       fileName={currentSong}
                       isPlaying={isPlaying} 
                       favoriteSongList={favoriteSongList.map(song => song.songName)}
                       onAddToFavorites={(songName) => onAddToFavorites(songName)} 
                       onRemoveFromFavorites={(songName) => onRemoveFromFavorites(songName)}
                       onStopPlaying={(fileName) => onStopPlaying(fileName)} 
                       onPlayNextSong={(fileName) => onPlayNextSong(fileName)} />
             <MusicUploadForm onAddSongToPlaylist={(event, songName, artistName, fileName) => onAddSongToPlaylist(event, songName, artistName, fileName)}
                              onUploadSong={onUploadSong} 
                              onError={(message) => onError(message)}
                              removeError={removeError}
                              />
        </div>
    );

}
