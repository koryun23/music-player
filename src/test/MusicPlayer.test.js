import renderer from "react-test-renderer"
import MusicPlayer from "../components/MusicPlayer";

it('renders correctly', () => {
    const tree = renderer.create(<MusicPlayer />).toJSON();
    console.log(tree);
    expect(tree).toMatchSnapshot();
})
