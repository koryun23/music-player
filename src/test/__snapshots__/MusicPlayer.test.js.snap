// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="music-player"
>
  <div
    className="container-fluid"
  >
    <div
      className="toolbar row"
    >
      <div
        className="col-md-2 col-sm-4 col-xs-12 custom-column"
      >
        <div
          className="dropdown dropdown-div play-all-button"
          onClick={[Function]}
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            className="button-description btn btn-light dropdown-toggle"
            data-toggle="dropdown"
            id="dropdownMenuButton"
            type="button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-play fa-lg play-icon"
              data-icon="play"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M73 39c-14.8-9.1-33.4-9.4-48.5-.9S0 62.6 0 80V432c0 17.4 9.4 33.4 24.5 41.9s33.7 8.1 48.5-.9L361 297c14.3-8.7 23-24.2 23-41s-8.7-32.2-23-41L73 39z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
            <b>
              Play All
            </b>
          </button>
          <div
            aria-labelledby="dropdownMenuButton"
            className="dropdown-menu"
          >
            <a
              className="dropdown-item"
              href="/"
              onClick={[Function]}
            >
              Repeat After Queue Ends
            </a>
            <a
              className="dropdown-item"
              href="/"
              onClick={[Function]}
            >
              Play Queue Only Once
            </a>
          </div>
        </div>
      </div>
      <div
        className="col-md-2 col-sm-4 col-xs-12 custom-column"
      >
        <div
          className="add-all-button dropdown dropdown-div"
          onClick={[Function]}
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            className="button-description btn btn-light dropdown-toggle"
            data-toggle="dropdown"
            id="dropdownMenuButton"
            type="button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-plus fa-lg add-icon"
              data-icon="plus"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
            <b>
              Add All
            </b>
          </button>
          <div
            className="dropdown-menu"
          >
            <a
              className="dropdown-item"
              href="/"
              onClick={[Function]}
            >
              Add All Sequentially
            </a>
            <a
              className="dropdown-item"
              href="/"
              onClick={[Function]}
            >
              Add All Shuffled
            </a>
          </div>
        </div>
      </div>
      <div
        className="col-md-6 col-sm-1 custom-column"
      />
      <div
        className="col-md-2 col-sm-3 col-xs-12 custom-column"
      >
        <form>
          <div
            className="filter-form"
          >
            <input
              className="filter-input-field"
              onChange={[Function]}
              placeholder="ï€‚ Filter"
              style={
                Object {
                  "fontFamily": "Arial, FontAwesome",
                }
              }
              type="text"
              value=""
            />
          </div>
        </form>
      </div>
      <div
        className="col-sm-1"
      />
    </div>
  </div>
  <table
    className="table table-hover"
  >
    <thead
      className="thead-light"
    >
      <tr>
        <th
          scope="col"
        />
        <th
          scope="col"
        >
          Song Name
        </th>
        <th
          scope="col"
        >
          Artist Name
        </th>
        <th
          scope="col"
        >
          Track
        </th>
        <th
          scope="col"
        />
        <th
          scope="col"
        />
        <th
          scope="col"
        />
      </tr>
    </thead>
    <tbody />
  </table>
  <form
    className="music-upload-form"
  >
    <div
      className="form form-group"
    >
      <input
        className="file-name"
        name="fileName"
        placeholder="No File Selected"
        readOnly={true}
        type="text"
        value=""
      />
      <small>
        Supported file formats: .mp3, .wav
      </small>
    </div>
    <div
      className="form form-group"
    >
      <input
        className="song-name"
        name="songName"
        onChange={[Function]}
        placeholder="Song Name"
        type="text"
        value=""
      />
    </div>
    <div
      className="form form-group"
    >
      <input
        className="artist-name"
        name="artistName"
        onChange={[Function]}
        placeholder="Artist Name"
        type="text"
        value=""
      />
    </div>
    <div
      className="form form-group"
    >
      <input
        className="btn btn-submit"
        name="submit"
        onChange={[Function]}
        type="file"
      />
      <button
        className="music-upload-button"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus fa-lg "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </form>
</div>
`;
